plugins/src/main/groovy/co/bound/codeartifact/CodeArtifactRepoProvider.groovy

```diff
@@ -1,53 +1,94 @@
 package co.bound.codeartifact
 
+import groovy.json.JsonOutput
+import groovy.json.JsonSlurper
 import org.gradle.api.artifacts.repositories.MavenArtifactRepository
+import org.gradle.api.credentials.Credentials
 import org.gradle.api.provider.Property
 import org.gradle.api.services.BuildService
 import org.gradle.api.services.BuildServiceParameters
+import org.gradle.internal.authentication.DefaultBasicAuthentication
+import org.gradle.internal.credentials.DefaultPasswordCredentials
 import software.amazon.awssdk.regions.Region
 import software.amazon.awssdk.services.codeartifact.CodeartifactClient
 import software.amazon.awssdk.services.codeartifact.model.GetAuthorizationTokenResponse
 
 import java.time.Instant
 
 abstract class CodeArtifactRepoProvider implements BuildService<Params> {
+
     private GetAuthorizationTokenResponse token = null
 
     interface Params extends BuildServiceParameters {
         Property<String> getDomain()
         Property<String> getAccountId()
         Property<String> getRegion()
         Property<String> getRepo()
+        Property<File> getGradleUserHome()
     }
 
     void configureRepo(MavenArtifactRepository spec) {
-        def params = getParameters()
-        String domain = getRequiredProperty(params.getDomain())
-        String accountId = getRequiredProperty(params.getAccountId())
-        String region = getRequiredProperty(params.getRegion())
-        String repo = getRequiredProperty(params.getRepo())
+        var params = getParameters()
+        var domain = getRequiredProperty(params.domain)
+        var accountId = getRequiredProperty(params.accountId)
+        var region = getRequiredProperty(params.region)
+        var repo = getRequiredProperty(params.repo)
 
         spec.name = "codeartifact"
         configureCodeArtifactUrl(spec, domain, accountId, region, repo)
-        spec.credentials {
-            it.username("aws")
-            it.password(this.getToken(domain, accountId, region))
+        spec.authentication {
+            add(new DefaultBasicAuthentication(spec.name) {
+                @Override
+                Credentials getCredentials() {
+                    return new DefaultPasswordCredentials("aws", "-") {
+                        @Override
+                        String getPassword() {
+                            return getToken(domain, accountId, region, params.gradleUserHome.get())
```

Comment:
Helps with lazy password evaluation

Reply:
Sorry, I didn't mean it will be lazy - it won't.

I was not able to find a way to make it lazy without using internal APIs. But given that we won't be calling a remote endpoint on every call to `getToken` now, leaving this eager is still an improvement. I meant that I'd rather leave it eager than start depending on internal APIs.

Reply:
Ok, removed usage of internal classes